modules/core/02-client/keeper/client_test.go

```diff
@@ -42,216 +42,186 @@ func (suite *KeeperTestSuite) TestCreateClient() {
+			// ensure fill height not set
+			_, found := suite.chainA.App.GetIBCKeeper().ClientKeeper.GetClientConsensusState(suite.chainA.GetContext(), path.EndpointA.ClientID, fillHeight)
+			suite.Require().False(found)
 
 			// updateHeader will fill in consensus state between prevConsState and suite.consState
 			// clientState should not be updated
-			updateHeader = createPastUpdateFn(suite)
-			return nil
+			updateHeader = createPastUpdateFn(fillHeight, trustedHeight)
 		}, true, false},
-		{"valid duplicate update", func() error {
-			clientState = ibctmtypes.NewClientState(testChainID, ibctmtypes.DefaultTrustLevel, trustingPeriod, ubdPeriod, maxClockDrift, testClientHeight, commitmenttypes.GetSDKSpecs(), ibctesting.UpgradePath, false, false)
-			clientID, err = suite.keeper.CreateClient(suite.ctx, clientState, suite.consensusState)
-			suite.Require().NoError(err)
+		{"valid duplicate update", func() {
+			clientID := path.EndpointA.ClientID
 
 			height1 := types.NewHeight(0, 1)
 
 			// store previous consensus state
 			prevConsState := &ibctmtypes.ConsensusState{
 				Timestamp:          suite.past,
-				NextValidatorsHash: suite.valSetHash,
+				NextValidatorsHash: suite.chainB.Vals.Hash(),
 			}
-			suite.keeper.SetClientConsensusState(suite.ctx, clientID, height1, prevConsState)
-
-			height2 := types.NewHeight(0, 2)
+			suite.chainA.App.GetIBCKeeper().ClientKeeper.SetClientConsensusState(suite.chainA.GetContext(), clientID, height1, prevConsState)
 
-			// store intermediate consensus state to check that trustedHeight does not need to be hightest consensus state before header height
-			intermediateConsState := &ibctmtypes.ConsensusState{
+			height5 := types.NewHeight(0, 5)
+			// store next consensus state to check that trustedHeight does not need to be hightest consensus state before header height
+			nextConsState := &ibctmtypes.ConsensusState{
 				Timestamp:          suite.past.Add(time.Minute),
-				NextValidatorsHash: suite.valSetHash,
+				NextValidatorsHash: suite.chainB.Vals.Hash(),
 			}
-			suite.keeper.SetClientConsensusState(suite.ctx, clientID, height2, intermediateConsState)
+			suite.chainA.App.GetIBCKeeper().ClientKeeper.SetClientConsensusState(suite.chainA.GetContext(), clientID, height5, nextConsState)
 
+			height3 := types.NewHeight(0, 3)
 			// updateHeader will fill in consensus state between prevConsState and suite.consState
 			// clientState should not be updated
-			updateHeader = createPastUpdateFn(suite)
+			updateHeader = createPastUpdateFn(height3, height1)
 			// set updateHeader's consensus state in store to create duplicate UpdateClient scenario
-			suite.keeper.SetClientConsensusState(suite.ctx, clientID, updateHeader.GetHeight(), updateHeader.ConsensusState())
-			return nil
+			suite.chainA.App.GetIBCKeeper().ClientKeeper.SetClientConsensusState(suite.chainA.GetContext(), clientID, updateHeader.GetHeight(), updateHeader.ConsensusState())
```

Comment:
I think in the future it is best to avoid setting all the consensus states manually, but this is fine. I think setting the update headers consensus state is fine, but the previous/intermediate consensus state (setup code) should just be done by calling `path.EndpointA.UpdateClient`. I think it helps in maintainability/readability in testing code

In my opinion, hard coding the heights is generally harder to maintain as opposed to referencing the latest height before/after an update.

Reply:
These are all past updates so it should be ok